// @nest-excalibur/common-api/lib/api/functions/search/complex-search/verificators-functions/verificator-helper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerificatorHelper = void 0;
const query_operators_1 = require("../constants/query-operators");
class VerificatorHelper {
    static verifyIsObject(value) {
        return value !== null && (typeof value === 'object') && !(value instanceof Array);
    }
    static verifyIsNotEmptyObject(object) {
        const keys = Object.keys(object);
        return keys.length > 0;
    }
    static isComplexOperatorObject(object) {
        const operatorKeys = Object.keys(object);
        const length = Object.keys(object).length;
        const hasOneKey = length === 2 || length === 1;
        if (hasOneKey) {
            const operator = operatorKeys[0];
            if (query_operators_1.QUERY_OPERATORS.includes(operator)) {
                const value = object[operator];
                if (typeof value === 'number' || typeof value === 'string' || (value && value.length >= 0)) {
                    if (length === 2) {
                        const conjunction = operatorKeys[1];
                        return conjunction === query_operators_1.OR_KEYWORD && typeof object[conjunction] === 'boolean';
                    }
                    return true;
                }
            }
        }
        return false;
    }
    static IsSimpleOr(object) {
        const hasProperties = object.conjunction && object.values;
        if (hasProperties) {
            if (object.conjunction === 'or' && object.conjunction === 'and') {
                return true;
            }
        }
        return false;
    }
    static isAttKeyWord(attr) {
        return query_operators_1.ATTR_KEYWORDS.includes(attr);
    }
}
exports.VerificatorHelper = VerificatorHelper;
