// @nest-excalibur/common-api/lib/api/decorators/custom-validator/is-type-or.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IsTypeOr = void 0;
const class_validator_1 = require("class-validator");
function IsTypeOr(validations, validationOptions) {
    return (object, propertyName) => {
        const validationsKeys = Object.keys(validations);
        class_validator_1.registerDecorator({
            name: 'wrongType',
            target: object.constructor,
            propertyName,
            options: validationOptions,
            validator: {
                validate(value, args) {
                    return validationsKeys.some(key => validations[key](value));
                },
                defaultMessage(validationArguments) {
                    const lastIndex = validationsKeys.length - 1;
                    const lastType = validationsKeys[lastIndex];
                    if (validationsKeys.length === 0)
                        return `Has to be ${lastType}`;
                    return `Can only be ${validationsKeys.join(', ')} or ${lastType}.`;
                }
            }
        });
    };
}
exports.IsTypeOr = IsTypeOr;
